src/goto-instrument/contracts/assigns.cpp

```diff
@@ -9,144 +9,177 @@ Date: July 2021
+  auto result = write_set.emplace(*this, target_expr);
 
-  if(!insertion_succeeded)
+  if(!result.second)
   {
     log.warning() << "Ignored duplicate expression '"
                   << from_expr(ns, target_expr.id(), target_expr)
                   << "' in assigns clause at "
                   << target_expr.source_location().as_string() << messaget::eom;
   }
+  return result.first;
 }
 
 void assigns_clauset::remove_from_write_set(const exprt &target_expr)
 {
-  write_set.erase(targett(*this, target_expr));
+  write_set.erase(conditional_address_ranget(*this, target_expr));
 }
 
-exprt assigns_clauset::targett::generate_containment_check(
-  const address_of_exprt &lhs_address) const
+exprt assigns_clauset::generate_inclusion_check(
+  const conditional_address_ranget &lhs) const
 {
-  const auto address_validity = and_exprt{
-    all_dereferences_are_valid(dereference_exprt{address}, parent.ns),
-    all_dereferences_are_valid(dereference_exprt{lhs_address}, parent.ns)};
-
-  exprt::operandst containment_check;
-  containment_check.push_back(same_object(lhs_address, address));
-
-  // If assigns target was a dereference, comparing objects is enough
-  // and the resulting condition will be:
-  // VALID(self) && VALID(lhs) ==> __CPROVER_same_object(lhs, self)
-  if(id != ID_dereference)
-  {
-    const auto lhs_offset = pointer_offset(lhs_address);
-    const auto own_offset = pointer_offset(address);
-
-    // __CPROVER_offset(lhs) >= __CPROVER_offset(target)
-    containment_check.push_back(
-      binary_relation_exprt(lhs_offset, ID_ge, own_offset));
-
-    const auto lhs_region = plus_exprt(
-      typecast_exprt::conditional_cast(
-        size_of_expr(lhs_address.object().type(), parent.ns).value(),
-        lhs_offset.type()),
-      lhs_offset);
-
-    const exprt own_region = plus_exprt(
-      typecast_exprt::conditional_cast(
-        size_of_expr(address.object().type(), parent.ns).value(),
-        own_offset.type()),
-      own_offset);
-
-    // (sizeof(lhs) + __CPROVER_offset(lhs)) <=
-    // (sizeof(self) + __CPROVER_offset(self))
-    containment_check.push_back(
-      binary_relation_exprt(lhs_region, ID_le, own_region));
-  }
-
-  // VALID(self) && VALID(lhs)
-  // ==> __CPROVER_same_object(lhs, self)
-  //  && __CPROVER_offset(lhs) >= __CPROVER_offset(self)
-  //  && (sizeof(lhs) + __CPROVER_offset(lhs)) <=
-  //        (sizeof(self) + __CPROVER_offset(self))
-  return or_exprt{not_exprt{address_validity}, conjunction(containment_check)};
-}
+  if(write_set.empty())
+    return not_exprt{lhs.validity_condition_var};
 
-goto_programt
-assigns_clauset::generate_havoc_code(const source_locationt &location) const
-{
-  modifiest modifies;
+  exprt::operandst conditions{not_exprt{lhs.validity_condition_var}};
   for(const auto &target : write_set)
-    modifies.insert(target.address.object());
+    conditions.push_back(target.generate_unsafe_inclusion_check(lhs));
```

Comment:
So you call it `unsafe` because the validity checks are inserted separately?

Reply:
Yes exactly.